### PR TITLE
feat(ivy): support templateUrl for ngtsc

### PR DESCRIPTION
**Pull request by alxhub**
This commit adds support for templateUrl in component templates within
ngtsc. The compilation pipeline is split into sync and async versions,
where asynchronous compilation invokes a special preanalyze() phase of
analysis. The preanalyze() phase can optionally return a Promise which
will delay compilation until it resolves.

A ResourceLoader interface is used to resolve templateUrls to template
strings and can return results either synchronously or asynchronously.
During sync compilation it is an error if the ResourceLoader returns a
Promise.

Two ResourceLoader implementations are provided. One uses 'fs' to read
resources directly from disk and is chosen if the CompilerHost doesn't
provide a readResource method. The other wraps the readResource method
from CompilerHost if it's provided.

### CODE REVIEW

**Comment by alfaproject** [comment]
Will this support TypeScript path resolution?

File: packages/compiler-cli/test/ngtsc/ngtsc_spec.ts [hunk 1]
```diff
@@ -147,6 +147,27 @@ describe('ngtsc behavioral tests', () => {
     expect(dtsContents).toContain('static ngComponentDef: i0.ComponentDef<TestCmp, \'test-cmp\'>');
   });
 
+  it('should compile Components without errors', () => {
```
Comment (mhevery) [inline packages/compiler-cli/test/ngtsc/ngtsc_spec.ts]: Can we have tests for the cases which throw errors.

**Review by mhevery — CHANGES_REQUESTED** [review]

File: packages/compiler-cli/test/ngtsc/ngtsc_spec.ts (hunk 1 above)
Comment (alxhub) [reply packages/compiler-cli/test/ngtsc/ngtsc_spec.ts]: Eventually we will. Right now the errors the compiler throws are temporary - basically assertions that `throw new Error(...)` on encountering something unexpected.

In the future, these will be migrated to become actual user-facing errors which will be reported via TypeScript's diagnostic system, including line/col pointers to the location of the error in the source. These are the errors we should write tests for.

**Review by mhevery — APPROVED** [review]

**Comment by mhevery** [comment]
http://test/OCL:203127686:BASE:203127689:1530632274073:dd5170b0

**Pull request closed without merging**